java-checks/src/test/java/org/sonar/java/checks/MissingPackageInfoCheckTest.java

```diff
@@ -0,0 +1,186 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.Assertions.in;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.sonar.java.checks.verifier.TestUtils.mainCodeSourcesPath;
+
+class MissingPackageInfoCheckTest {
+
+  private static final String EXPECTED_PACKAGE = "checks.packageInfo.nopackageinfo";
+  private static final String EXPECTED_MESSAGE = "Add a 'package-info.java' file to document the '" + EXPECTED_PACKAGE + "' package";
+
+  @RegisterExtension
+  public final LogTesterJUnit5 logTester = new LogTesterJUnit5();
+
+  private ReadCache readCache;
+  private InternalWriteCache writeCache;
+  private CheckVerifier verifier;
+
+  @BeforeEach
+  void initVerifier() {
+    this.readCache = new InternalReadCache();
+    this.writeCache = new InternalWriteCache().bind(readCache);
+    this.verifier = CheckVerifier.newVerifier()
+      .withCache(readCache, writeCache);
+  }
+
+  @Test
+  void no_package_info() {
+    MissingPackageInfoCheck check = new MissingPackageInfoCheck();
+
+    CheckVerifier.newVerifier()
+      .onFiles(
+        mainCodeSourcesPath("DefaultPackage.java"),
+        mainCodeSourcesPath("checks/packageInfo/HelloWorld.java"),
+        mainCodeSourcesPath("checks/packageInfo/package-info.java"),
+        mainCodeSourcesPath("checks/packageInfo/nopackageinfo/HelloWorld.java"),
+        mainCodeSourcesPath("checks/packageInfo/nopackageinfo/nopackageinfo.java"))
+      .withCheck(check)
+      .verifyIssueOnProject(EXPECTED_MESSAGE);
+
+    Set<String> set = check.missingPackageWithoutPackageFile;
+    assertThat(set).hasSize(1);
+    assertThat(set.iterator().next()).isEqualTo(EXPECTED_PACKAGE);
+  }
+
+  @Test
+  void caching() {
+    verifier
+      .onFiles(
+        mainCodeSourcesPath("DefaultPackage.java"),
+        mainCodeSourcesPath("checks/packageInfo/HelloWorld.java"),
+        mainCodeSourcesPath("checks/packageInfo/package-info.java"),
+        mainCodeSourcesPath("checks/packageInfo/nopackageinfo/HelloWorld.java"),
+        mainCodeSourcesPath("checks/packageInfo/nopackageinfo/nopackageinfo.java")
+      )
+      .withCheck(new MissingPackageInfoCheck())
+      .verifyIssueOnProject(EXPECTED_MESSAGE);
+
+    var check = spy(new MissingPackageInfoCheck() {
+      @Override
+      public boolean scanWithoutParsing(InputFileScannerContext inputFileScannerContext) {
+        this.context = inputFileScannerContext;
+        return super.scanWithoutParsing(inputFileScannerContext);
+      }
+
+      @Override
+      public void scanFile(JavaFileScannerContext context) {
+        this.context = context;
+        super.scanFile(context);
+      }
+    });
```

Comment:
It's strange to override the class that you want to test. How could you be sure that the class will behave the same without those overrides?
And the class `MissingPackageInfoCheck` receives `InputFileScannerContext context` as a method parameter for `processFile` and as a field for `endOfAnalysis`.
Why not add a `setContext` method that will be also used in the test, and change `processFile` to also depends on the field `context`.